dev/core#111 Support Custom Data for MembershipType entity

mattwire opened the pull request:
Overview
----------------------------------------
Enable custom data support on the MembershipType form.

Before
----------------------------------------
Could not view/edit custom data on MembershipType form.

After
----------------------------------------
Can view/edit custom data on MembershipType form.

Technical Details
----------------------------------------
- Add generic helper for setDeleteMessage() in EntityFormTrait so we can simplify translation (one string for all forms instead of one for each form).
- Change CRM/Member/Form/MembershipType.php to use MembershipType.create API instead of BAO add (so we can save custom data).
- Change domain_id to non-mandatory for MembershipType.create API and default to current domain (see https://lab.civicrm.org/dev/core/issues/113 for discussion).
- Convert MembershipType form to use EntityFormTrait so we automatically add custom data to the UI and use more generic code.

Comments
----------------------------------------
@eileenmcnaughton This is a follow on from all the previous ones (eg. #12117, #12118).  It should address all the previous issues.  Also note that I'm now using cg_extends to add MembershipType (you'll either need to add it manually, or install this extension: https://github.com/mattwire/uk.co.mjwconsult.variablerecurpayments
Which will setup cg_extends for you as well as provide some handy MembershipType custom fields for testing :-)

mattwire commented:
test this please

mattwire commented:
test this please

eileenmcnaughton commented on api/v3/MembershipType.php:
```diff
@@ -47,6 +47,10 @@ function civicrm_api3_membership_type_create($params) {
   }
+
+  if (!isset($params['domain_id'])) {
+    $params['domain_id'] = CRM_Core_Config::domainID();
```
This would incorrectly be set if 'id' is set - I think we should spin this out & resolve it first - but I think we should support 'current_domain' in the same way we support 'user_contact_id', now - ie. ie would look like

$params['domain_id']['api.default'] => 'current_domain_id';

I would need to understand from @totten why the code in utils handles '('or similar') but I guess we could just 'special-handle' the specific string 'current_domain_id'


  if (!empty($fieldValue) || $fieldValue === '0' || $fieldValue === 0) {
    // if value = 'user_contact_id' (or similar), replace value with contact id
    if (!is_numeric($fieldValue) && is_scalar($fieldValue)) {
      $realContactId = _civicrm_api3_resolve_contactID($fieldValue);
      if ('unknown-user' === $realContactId) {
        throw new API_Exception("\"$fieldName\" \"{$fieldValue}\" cannot be resolved to a contact ID", 2002, array('error_field' => $fieldName, "type" => "integer"));
      }

eileenmcnaughton commented on CRM/Admin/Form/RelationshipType.php:
```diff
@@ -97,7 +97,7 @@ public function getDefaultEntity() {
    */
   public function setDeleteMessage() {
-    $this->deleteMessage = ts('WARNING: Deleting this option will result in the loss of all Relationship records of this type.') . ts('This may mean the loss of a substantial amount of data, and the action cannot be undone.') . ts('Do you want to continue?');
+    self::setEntityDeleteMessage(ts('Relationship'));
```
@mlutfy can you comment on whether this interpolation is equally good from your point of view? I originally assumed that the way it was before (with a line for each entity) was 'better' but not so sure now...

mattwire replied:
I started changing it separately for each entity but then realised that nearly every entity with a delete action shares exactly the same wording (except the entity name) so I think this will reduce the number of strings that need translating significantly.

eileenmcnaughton replied:
So - I'm not sure if it is clear without the world - I'm not inclined to make a change with a negative impact on usability for non-anglophones where there is such a minor code repetition decrease

mattwire commented:
@eileenmcnaughton This is now updated and rebased.  I removed the "translation" helper for setDeleteMessage() I realise it doesn't really help (per @mlutfy comments).  Also removed the domain_id as MembershipType API has already been updated in core to make this optional.  Pending tests passing I think this should be ok to merge.

eileenmcnaughton commented:
@mattwire I'm not ready to accept that we can't use the entity fields.tpl yet - I want to try to push a bit further on that before accepting it. The main reason is that I want to set a bar for adding custom fields to an entity. I'm happy to merge this without the one line that delivers the main functionality - {include file="CRM/common/customDataBlock.tpl"} and then take a look at how we can go further on this

eileenmcnaughton commented:
Per previous comments - adding merge on pass - I'll take a look & see if we can do anything with the tpl

eileenmcnaughton commented:
test this please
